### PR TITLE
Rework session helpers

### PR DESCRIPTION
**Pull request by PaulAsjes**
## Description

Refactors the session helpers to add a new session class. The old helper methods will stay put, but we'll likely add a deprecation comment and remove them in a future major version.

See this [gist](https://gist.github.com/PaulAsjes/e1f80d5451f988db2acdf6ee894fcc46) as an example of how this would work in practise. 

## Documentation

Does this require changes to the WorkOS Docs? E.g. the [API Reference](https://workos.com/docs/reference) or code snippets need updates.

```
[x] Yes
```

If yes, link a related docs PR and add a docs maintainer as a reviewer. Their approval is required.

### CODE REVIEW

File: src/user-management/interfaces/authenticate-with-session-cookie.interface.ts [hunk 1]
```diff
@@ -34,4 +36,6 @@ export type AuthenticateWithSessionCookieSuccessResponse = {
   organizationId?: string;
   role?: string;
   permissions?: string[];
+  user?: User;
```
Comment (mattgd) [inline src/user-management/interfaces/authenticate-with-session-cookie.interface.ts:39]: Is user guaranteed on successful auth, or no?

File: src/user-management/user-management.ts [hunk 2]
```diff
@@ -511,7 +524,7 @@ export class UserManagement {
     authenticationResponse: AuthenticationResponse;
     session?: AuthenticateWithSessionOptions;
   }): Promise<AuthenticationResponse> {
-    if (session) {
+    if (session?.sealSession) {
```
Comment (mattgd) [inline src/user-management/user-management.ts:527]: 🙌

File: src/user-management/user-management.ts [hunk 3]
```diff
@@ -155,19 +156,28 @@ const toQueryString = (options: Record<string, string | undefined>): string => {
...
+
     // Set the JWKS URL. This is used to verify if the JWT is still valid
     this.jwks = clientId
       ? createRemoteJWKSet(new URL(this.getJwksUrl(clientId)))
       : undefined;
   }
 
+  session(sessionData: string, cookiePassword: string): Session {
+    return new Session(this, sessionData, cookiePassword);
+  }
```
Comment (mattgd) [inline src/user-management/user-management.ts:179]: I wonder if this should be called something like `sessionFromCookie()`. Also do we want to allow `WORKOS_COOKIE_PASSWORD` from environment variable at all? We currently do this, but I don't have a strong opinion either way.

File: src/user-management/user-management.ts [hunk 4]
```diff
@@ -484,7 +494,10 @@ export class UserManagement {
         };
       }
 
-      return { authenticated: true, sealedSession };
+      return {
+        authenticated: true,
+        session: sealedSession,
+      };
```
Comment (mattgd) [inline src/user-management/user-management.ts:500]: Breaking change?

File: src/user-management/session.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+          organizationId:
+            options.organizationId ?? organizationIdFromAccessToken,
+          session: {
+            // We want to store the new sealed session in this class instance, so this always needs to be true
+            sealSession: true,
+            cookiePassword: options.cookiePassword ?? this.cookiePassword,
+          },
+        });
+
+      if (options.sealSession) this.cookiePassword = options.cookiePassword;
```
Comment (mattgd) [inline src/user-management/session.ts:149]: I'm strongly pro-curly braces for maintainability.
https://www.imperialviolet.org/2014/02/22/applebug.html
```suggestion
      if (options.sealSession) {
        this.cookiePassword = options.cookiePassword;
      }
```

File: src/user-management/interfaces/refresh-and-seal-session-data.interface.ts [hunk 6]
```diff
@@ -16,14 +18,15 @@ export enum RefreshAndSealSessionDataFailureReason {
...
+// TODO: These should be renamed since it's possible to have an unsealed session
 type RefreshAndSealSessionDataFailedResponse = {
   authenticated: false;
   reason: RefreshAndSealSessionDataFailureReason;
 };
 
 type RefreshAndSealSessionDataSuccessResponse = {
   authenticated: true;
-  sealedSession: string;
+  session: string | AuthenticationResponse;
```
Comment (mattgd) [inline src/user-management/interfaces/refresh-and-seal-session-data.interface.ts:29]: Breaking change?

File: src/user-management/user-management.ts (hunk 3 above)
Comment (PaulAsjes) [reply src/user-management/user-management.ts:179]: I kinda like the terseness of `session()`.

I want to shy away from using environment variables as it causes headaches in environments that don't support `process` (looking at you CFW).

File: src/user-management/interfaces/refresh-and-seal-session-data.interface.ts (hunk 6 above)
Comment (PaulAsjes) [reply src/user-management/interfaces/refresh-and-seal-session-data.interface.ts:29]: Bah, you're right. I think we'll need a `sealedSession` and `session` to not break this.

File: src/user-management/user-management.ts (hunk 3 above)
Comment (mattgd) [reply src/user-management/user-management.ts:179]: Good point on CFW, works for me!

File: src/user-management/interfaces/authenticate-with-session-cookie.interface.ts (hunk 1 above)
Comment (PaulAsjes) [reply src/user-management/interfaces/authenticate-with-session-cookie.interface.ts:39]: Should do, can't think of a scenario with successfully authenticating doesn't get you the user object.

**Review by mattgd — APPROVED** [review]
Looks good, thanks for updating! Left one follow up comment.

**Pull request merged**